src/federation.cpp

```diff
@@ -0,0 +1,843 @@
+#include <node/context.h>
+#include <pow.h>
+#include <primitives/block.h>
+#include <pubkey.h>
+#include <script/script.h>
+#include <script/standard.h>
+#include <streams.h>
+#include <sync.h>
+#include <threadinterrupt.h>
+#include <util/strencodings.h>
+#include <util/system.h>
+#include <validation.h>
+#ifdef ENABLE_WALLET
+#include <script/generic.hpp>
+#include <script/sign.h>
+#include <wallet/wallet.h>
+#endif // ENABLE_WALLET
+
+#include <algorithm>
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <map>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+//! Pointer to the NodeContext for the process
+static NodeContext* g_context = nullptr;
+
+//! Critical section guarding access to any of the block-signing federation global state
+static RecursiveMutex cs_block_signer;
+
+//! Thread interrupter which is used to command the block-signing thread to terminate.
+static CThreadInterrupt g_thread_interrupt;
+void InterruptBlockSignerThread() {
+    g_thread_interrupt();
+}
+
+#ifdef ENABLE_WALLET
+// Elements supports multiple user wallets.  The wallet which should be used for
+// block signing can be specified via a configuration option, otherwise the
+// "default" wallet will be used.
+static std::shared_ptr<CWallet> GetWalletForBlockSigning()
+{
+    std::vector<std::shared_ptr<CWallet>> wallets = GetWallets();
+    if (wallets.empty()) {
+        return nullptr;
+    }
+    // The user can configure which wallet to use for block signing with the
+    // '-signblockwallet' option.  By default, the default wallet is used.
+    const std::string requestedwallet = g_context->args->GetArg("-signblockwallet", "");
+    std::shared_ptr<CWallet> ret = GetWallet(requestedwallet);
+    if (ret) {
+        return ret;
+    }
+    // If a wallet name was specified but not found, let's log that so the user
+    // knows they need to fix their configuration.
+    if (requestedwallet != "" && requestedwallet != "0") {
+        LogPrintf("Requested wallet \"%s\" be used to sign block proposals, but no such wallet found.\n", requestedwallet);
+    }
+    // The user can disable use of a wallet for block signing by setting
+    // '-signblockwallet' to 0 or false or '-nosignblockwallet=1'.  Note that
+    // there must not be a wallet named "0" (or "false", etc.), or else that
+    // wallet will be selected above.
+    if (!g_context->args->GetBoolArg("-signblockwallet", true)) {
+        return nullptr;
+    }
+    // If we get this far, it is because the default wallet is requested.
+    return !wallets.empty() ? wallets[0] : nullptr;
+}
+#endif // ENABLE_WALLET
+
+//! All public keys contained within the block signing script.
+static std::vector<CPubKey> g_block_signing_keys;
+//! The block signing public key associated with this node.
+static CPubKey g_our_block_signing_key;
+//! The index of the block signing key associated with this node.
+static int g_our_block_signing_key_index = 0;
```

Comment:
Might be a good idea to switch this to a `size_t` since you're not relying on any -1==not_found things AFAICT. It also fixes
```
  CXX      libbitcoin_server_a-federation.o
federation.cpp:166:39: warning: comparison of integers of different signs: 'int' and 'std::__1::vector<CPubKey, std::__1::allocator<CPubKey> >::size_type' (aka 'unsigned long') [-Wsign-compare]
    if (g_our_block_signing_key_index >= g_block_signing_keys.size()) {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Done.

And I guess I need to upgrade clang on my dev machine, because it's weird I didn't get those warnings.